### ISSUE TITLE
Rust-only release

### ISSUE THREAD

**Issue by MilanVasko** [gltf-rs/mikktspace#16]
Hello!
I have been playing around with the [Amethyst](https://amethyst.rs/) engine, which uses `mikktspace 0.1` in its GLTF importer. When I was trying to run it on Windows 10, it failed during compilation with this error:

```
error: linking with `link.exe` failed: exit code: 1120
  |
  = note: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.24.28314\\bin\\HostX64\\x86\\link.exe" "/NOLOGO" "/NXCOMPAT" ... line too long to paste here
  = note: libmikktspace-ed048d53099c6187.rlib(mikktspace-ed048d53099c6187.mikktspace.24tkbvwf-cgu.0.rcgu.o) : error LNK2019: unresolved external symbol _genTangSpaceDefault@4 referenced in function __ZN10mikktspace8Closures8generate17hfe0598a16c7e4f72E
          C:\Users\mvask\Projects\amethyst\target\debug\examples\gltf.exe : fatal error LNK1120: 1 unresolved externals
```

It seems like there was a problem while compiling the C part of this library.

I've tried to update the dependency on `mikktspace` to the latest commit on master, since I've understood that since 0.1, there's been some work done to port the C parts of the library to pure Rust - which resolved my issue.

The problem with that approach is that it forces the use of a git dependency, since the last release of `mikktspace` I've found still has some parts written in C. [As I've learned](/amethyst/amethyst/pull/2085), using a git dependency has some drawbacks, which ultimately brings me here with a question:

Can a new, Rust-only version of mikktspace be released on crates.io? If not, what needs to be done?

**Comment by alteous** [comment]
> Can a new, Rust-only version of mikktspace be released on crates.io? If not, what needs to be done?

There is an unreleased Rust-only version in [src/generated.rs](https://github.com/gltf-rs/mikktspace/blob/master/src/generated.rs). It's unreleased because it has hardly been tested. I'd consider releasing under an 'experimental' feature tag in the meantime.

**Comment by MilanVasko** [comment]
> I'd consider releasing under an 'experimental' feature tag in the meantime.

That sounds good, thanks!

> It's unreleased because it has hardly been tested.

Is there a way I can help with that?

**Comment by azriel91** [comment]
Heya, I opened #17 to attempt to allow for the experimental release. Since there seems to be one API entry point (besides the trait), I just made it panic if the `"experimental"` feature isn't enabled.

Right now the panic message doesn't include the version, since I'm not sure what version this will be released under (`0.2.0-alpha`?).

**Comment by alteous** [comment]
> Is there a way I can help with that?

I'd like to have some kind of regression test to show whether generated tangents differ from a reference version.

**Comment by MilanVasko** [comment]
> I'd like to have some kind of regression test to show whether generated tangents differ from a reference version.

Allright, I'll see what I can do.

**Comment by alteous** [comment]
When appropriate, we should add testing to CI for both the C and Rust backends.

**Comment by MilanVasko** [comment]
> When appropriate, we should add testing to CI for both the C and Rust backends.

Sorry, I am a bit confused now. Looking at the [output of Travis](https://travis-ci.org/gltf-rs/mikktspace/jobs/637586823), it seems the test is already integrated to the pipeline. Also, in master, there is currently no C code at all.

**Comment by alteous** [comment]
Sorry I completely forgot the C code was removed! Ignore my previous comments. :man_facepalming:

In that case there's no point in releasing under an `experimental` feature flag either.

**Comment by alteous** [comment]
@MilanVasko, to confirm: the reference data added in #19 was generated by the original C code? If that's so then I'm happy to release as-is.

**Comment by MilanVasko** [comment]
> @MilanVasko, to confirm: the reference data added in #19 was generated by the original C code? If that's so then I'm happy to release as-is.

Yes, it was :) Feel free to check the implementation of the generator, I've published it [here](/MilanVasko/mikktspace-test-helper). The actual code that generates the data is [here](/MilanVasko/mikktspace-test-helper/blob/master/src/main.cpp#L283).

**Comment by MilanVasko** [comment]
@alteous Hello, any updates? :smile:

**Comment by alteous** [comment]
Sorry for the lack of response. I was on the cusp of releasing but then I had reservations about the design and how it might be misunderstood. In the spirit of worse-is-better and to unblock people I see little harm in releasing as-is and so I'll do so today.

My reservations were:

1. Implementing the `Geometry` trait for a `struct` (or otherwise) might mislead people to believe the `set_tangent*` functions set the tangent for the current set of geometry. There is a subtle difference: the output geometry is actually re-ordered and may need to be 'welded' back together, see http://www.mikktspace.com for the details.

2. The `Geometry::set_tangent_encoded` doesn't work if the user implements both `set_tangent` and `set_tangent_encoded`. This is a minor issue really and could be explained in the documentation.

3. I thought the name `generate_tangents_default` was a bit odd since there are no other alternatives than the default! I considered adding a `AngularThreshold` struct (with `Default` impl = 180.0) so the function would look like this instead: `fn generate_tangents<G: Geometry>(geometry: &mut G, threshold: AngularThreshold) -> bool { ... }`. I decided against this eventually since I couldn't find an explanation of what the angular threshold actually does. It seems to be a feature for debugging.

**Comment by alteous** [comment]
Released as version 0.2.0
